File: nativeLib/rary.meteorological/src/dgeocomps.f
 
      Subroutine DgeoComps(Z,F,
     -                     SpaX,SpaY,mnx,mny,nx,ny,
     -                     dugdx,dugdy,dvgdx,dvgdy)

C  Z(mnx,mny)       Real   Input height field.
C  F(mnx,mny)       Real   Array of coriolis parameter.
C  SpaX(mnx,mny)    Real   Grid spacing in the X direction.
C  SpaY(mnx,mny)    Real   Grid spacing in the Y direction.
C  mnx,mny          Int    Dimensions of input arrays.
C  nx,ny            Int    Dimensions of input grids.
C  dugdx(mnx,mny)   Real   d/dx of u component of geostrophic wind.
C  dugdy(mnx,mny)   Real   d/dy of u component of geostrophic wind.
C  dvgdx(mnx,mny)   Real   d/dx of v component of geostrophic wind.
C  dvgdy(mnx,mny)   Real   d/dy of v component of geostrophic wind.
 
 
      Implicit None
 
      Integer*4 mnx,mny,nx,ny,im,ip,jm,jp,i,j
 
      Real*4    dugdx(mnx,mny),dugdy(mnx,mny),
     -          dvgdx(mnx,mny),dvgdy(mnx,mny),
     -          F(mnx,mny),SpaX(mnx,mny),SpaY(mnx,mny),
     -          Z(mnx,mny),
     -          g,qqq,www

      Real*4 Flag,Flg
c      Common/FlagFlg/Flag,Flg
      Data Flag,Flg/1e37,99998.0/
      Data g/9.806/

c  Check validity of input parameters.
      If (nx.lt.3 .or. nx.gt.mnx .or. ny.lt.3 .or. ny.gt.mny) Return

c  Flag out boundaries.
      Do 21 i=1,nx
21    dugdx(i,1)=Flag
       Do 22 j=2,ny-1
       dugdx(1,j)=Flag
22     dugdx(nx,j)=Flag
        Do 23 i=1,nx
23      dugdx(i,ny)=Flag
      Do 31 i=1,nx
31    dugdy(i,1)=Flag
       Do 32 j=2,ny-1
       dugdy(1,j)=Flag
32     dugdy(nx,j)=Flag
        Do 33 i=1,nx
33      dugdy(i,ny)=Flag
      Do 41 i=1,nx
41    dvgdx(i,1)=Flag
       Do 42 j=2,ny-1
       dvgdx(1,j)=Flag
42     dvgdx(nx,j)=Flag
        Do 43 i=1,nx
43      dvgdx(i,ny)=Flag
      Do 51 i=1,nx
51    dvgdy(i,1)=Flag
       Do 52 j=2,ny-1
       dvgdy(1,j)=Flag
52     dvgdy(nx,j)=Flag
        Do 55 i=1,nx
55      dvgdy(i,ny)=Flag

c  Calculate components.
3000  jm=1
      j=2
      Do 3091 jp=3,ny
      im=1
      i=2
      Do 3090 ip=3,nx
      If (Z(im,jm).gt.Flg .or. Z(im,jp).gt.Flg .or.
     -    Z(ip,jm).gt.Flg .or. Z(ip,jp).gt.Flg .or.
     -    Z(i,jm).gt.Flg .or. Z(i,jp).gt.Flg .or.
     -    Z(im,j).gt.Flg .or. Z(ip,j).gt.Flg .or.
     -    Z(i,j).gt.Flg) Then
          dugdx(i,j)=Flag
          dugdy(i,j)=Flag
          dvgdx(i,j)=Flag
          dvgdy(i,j)=Flag
      Else
          qqq=g/F(i,j)
          www=Z(i,j)+Z(i,j)
          dvgdy(i,j)=qqq*(Z(ip,jp)-Z(ip,jm)-Z(im,jp)+Z(im,jm))/
     -               (4*SpaX(i,j)*SpaY(i,j))
          dugdx(i,j)=-dvgdy(i,j)
          dugdy(i,j)=qqq*(www-Z(i,jp)-Z(i,jm))/(SpaY(i,j)*SpaY(i,j))
          dvgdx(i,j)=qqq*(Z(ip,j)+Z(im,j)-www)/(SpaX(i,j)*SpaX(i,j))
      End If
      im=i
      i=ip
3090  Continue
      jm=j
      j=jp
3091  Continue

      Return
      End

File: edexOsgi/com.raytheon.uf.common.derivparam.python/utility/common_static/base/derivedParameters/functions/DgeoComps.py
##
# This software was developed and / or modified by Raytheon Company,
# pursuant to Contract DG133W-05-CQ-1067 with the US Government.
# 
# U.S. EXPORT CONTROLLED TECHNICAL DATA
# This software product contains export-restricted data whose
# export/transfer/disclosure is restricted by U.S. law. Dissemination
# to non-U.S. persons whether in the United States or abroad requires
# an export license or other authorization.
# 
# Contractor Name:        Raytheon Company
# Contractor Address:     6825 Pine Street, Suite 340
#                         Mail Stop B8
#                         Omaha, NE 68106
#                         402.291.0100
# 
# See the AWIPS II Master Rights File ("Master Rights File.pdf") for
# further licensing information.
###

## @file DgeoComps.py

from numpy import isscalar
from numpy import copy as Copy
from numpy import NaN
from numpy import ones
from numpy import shape

##
# Compute the components of geostrophic wind for this level.
# @attention: result may contain NaN.
#
# @param height: Height (m)
# @type height: 2D numpy array, at least 3x3
# @param dx: Spacing between data points in the X direction
# @type dx: scalar or 2D numpy array at least 3x3
# @param dy: Spacing between data points in the Y direction
# @type dy: scalar or 2D numpy array at least 3x3
# @param coriolis: Coriolis force (kg-m/s^2)
# @type coriolis: scalar or 2D numpy array at least 3x3
# @return: d/dx of u component of geostrophic wind
#          d/dy of u component of geostrophic wind
#          d/dx of v component of geostrophic wind
#          d/dy of v component of geostrophic wind
# @rtype: tuple(dugdx, dugdy, dvgdx, dvgdy)
def execute(height, dx, dy, coriolis):
    "Compute the components of geostrophic wind for this level."
    # Generate output arrays as arrays of NaN,
    # the same size and dtype as Height.
    # dugdx = masked_where(True, Height) -- not needed
    dugdy = ones(shape(height), dtype=height.dtype) * NaN 
    dvgdx = Copy(dugdy)
    dvgdy = Copy(dugdy)
    
    gravAcc = 9.806
    
    # use cropped versions of dx, dy, and coriolis in calculations
    if not isscalar(dx):
        dx = dx[1:-1,1:-1]

    if not isscalar(dy):
        dy = dy[1:-1,1:-1]

    if not isscalar(coriolis):
        coriolis = coriolis[1:-1,1:-1]
    
    # intermediate values that are used repeatedly
    qqq = gravAcc / coriolis
    www = height[1:-1,1:-1] * 2.0

    # make calculations in arrays with outer edge cropped
    # then paste into full-sized array
    croppedDvgdy = height[2:,2:] + height[0:-2,0:-2]
    croppedDvgdy -= height[2:,0:-2] + height[0:-2,2:]
    croppedDvgdy *= qqq
    croppedDvgdy /= 4 * dx * dy
    dvgdy[1:-1,1:-1] = croppedDvgdy
    
    dugdx = -dvgdy#*0.0
    
    croppedDugdy = www - height[2:,1:-1] - height[0:-2,1:-1]
    croppedDugdy *= qqq
    croppedDugdy /= dy * dy
    dugdy[1:-1,1:-1] = croppedDugdy
    
    croppedDvgdx = height[1:-1,2:] + height[1:-1,0:-2] - www
    croppedDvgdx *= qqq
    croppedDvgdx /= dx * dx
    dvgdx[1:-1,1:-1] = croppedDvgdx
    
    return (dugdx, dugdy, dvgdx, dvgdy)